pkg/reconcilermanager/controllers/reposync_controller.go

```diff
@@ -198,113 +288,228 @@ func (r *RepoSyncReconciler) Reconcile(ctx context.Context, req controllerruntim
-	case statusCurrent:
+		reposync.SetStalled(rs, "Deployment", errors.New(string(result.Status)))
+	case kstatus.CurrentStatus:
 		// currentStatus indicates that the deployment is available, which qualifies
 		// to clear the Reconciling status condition in RepoSync.
 		reposync.ClearCondition(rs, v1beta1.RepoSyncReconciling)
 		// Since there were no errors, we can clear any previous Stalled condition.
 		reposync.ClearCondition(rs, v1beta1.RepoSyncStalled)
+	default:
+		return errors.Errorf("invalid deployment status: %v: %s", result.Status, result.Message)
 	}
 
 	updated, err := r.updateStatus(ctx, currentRS, rs)
-	// Use the status update error for metric tagging, if no other errors.
-	metrics.RecordReconcileDuration(ctx, metrics.StatusTagKey(err), start)
 	if err != nil {
-		return controllerruntime.Result{}, err
+		return err
 	}
 
-	if updated && result.status == statusCurrent {
-		r.log.Info("Deployment successfully reconciled", operationSubjectName, reconcilerName, executedOperation, op)
+	if updated && result.Status == kstatus.CurrentStatus {
+		log.Info("Deployment fully available", executedOperation, op)
 	}
-	return controllerruntime.Result{}, nil
+	return nil
+}
+
+// deleteManagedObjects deletes objects managed by the reconciler-manager for
+// this RepoSync. If updateStatus is true, the RepoSync status will be updated
+// to reflect the teardown progress.
+func (r *RepoSyncReconciler) deleteManagedObjects(ctx context.Context, log logr.Logger, reconcilerKey types.NamespacedName, currentRS, rs *v1beta1.RepoSync, updateStatus bool) error {
+	log = log.WithValues("reconciler", reconcilerKey.String())
+	log.Info("Deleting managed resources")
+
+	if updateStatus {
+		// If no Reconciling condition is not already Termination, set it to Termination.
+		// Avoid updating the timestamps on every retry.
+		cond := reposync.GetCondition(rs.Status.Conditions, v1beta1.RepoSyncReconciling)
+		if cond == nil || cond.Reason != "Termination" {
+			reposync.SetReconciling(rs, "Termination", "Teardown in progress")
+			if _, err := r.updateStatus(ctx, currentRS, rs); err != nil {
+				return err
+			}
+		}
+	}
+
+	err := r.deleteDeployment(ctx, reconcilerKey)
+	if err != nil {
+		log.Error(err, "RepoSync Deployment delete failed")
+		if updateStatus {
+			reposync.SetStalled(rs, "Deployment", err)
+			// Delete errors should always trigger retry (return error),
+			// even if status update is successful.
+			if _, updateErr := r.updateStatus(ctx, currentRS, rs); updateErr != nil {
+				log.Error(updateErr, "RepoSync status update failed")
+			}
+		}
+		// Return the upsertDeployment error, not the updateStatus error.
+		return errors.Wrap(err, "RepoSync Deployment delete failed")
+	}
+
+	// Note: ConfigMaps have been replaced by Deployment env vars.
+	// Using env vars auto-updates the Deployment when they change.
+	// This deletion remains to clean up after users upgrade.
+
+	if err := r.deleteConfigmap(ctx, reconcilerKey); err != nil {
```

Comment:
Looks like each of these `delete*` are using the same control flow. WDYT about creating an array of these functions/resources and looping through them?

Reply:
I've been noticing that a lot of the upsert and delete funcs are very similar. I think it makes sense to refactor, but I'm a little scared on making too many changes here, since it will require changing the tests more. But it seems like a good follow up.